How many late days are you using for this assignment? 

In [ ]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [82]:
# Load the given datasets
cat = pd.read_csv("chronic_kidney_disease_categorical.csv")
num = pd.read_csv("chronic_kidney_disease_numerical.csv")
# Print the data
print(cat)
print(num)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and  caterical columns each (if they exist)

Answer: The data is medical data used to predict the presence of CKD in patients. The dataset includes various attributes (features) that describe patient health conditions. The features are the varibles within the CSV and are what will end up being the columns of the datset. The features range from  caterail, to categorical, binary, continous. sg, al , su are categroical while age bu and bgr are integers. The label is a class indicanting wehtehr the patient has ckd or nockd. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [83]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
if int(num.duplicated(subset="unique_id", keep="first").sum()) > 0:
    print(num[num.duplicated(subset="unique_id", keep="first")])
    num = num.drop_duplicates(subset="unique_id", keep="first")

if int( cat.duplicated(subset="unique_id", keep="first").sum()) > 0:
    print(cat[cat.duplicated(subset="unique_id", keep="first")])
    cat = cat.drop_duplicates(subset="unique_id", keep="first")


# # Drop these duplicate rows

# # Repeat the same for categorical dataset. Print the duplicate rows and drop them


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
107     872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
149     721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
171     240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
173     441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
185     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
194     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  46.0   
213     354223  64.0   70.0   97.0   27.0  0.7  145.0  4.8  13.8  49.0   
244     302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
245     431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
248     687021  52.0   80.0   99.0   25.0  0.8  135.0  3.7  15.0  52.0   
249     843449  30.0   80.0  131.0   38.0  1.0  147.0  3.8  14.1  45.0   
265     628271  65.0   90.0  270.0   39.0  2.0    NaN  NaN  12.0  36.0   
273     862719  61.0   80.0  131.0   2

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table? 
No not all unquie ids are present in both. I know this from doing a comparrsion of the two by seeing if any unique id was not in or the other. I found that in the following comparssionn in the nums_dataset that it is missing the folloinwg 208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335. In the coombined table the rows that are missing keep there data from categorical but any of the data for any of the numercail data columns they are filled with NAN. 

Answer: 

In [84]:
ids_num = set(num["unique_id"])
ids_cat = set(cat["unique_id"])


missing_in_numerical = ids_num - ids_cat  
missing_in_categorical = ids_cat - ids_num 

print("elements in ids_num but not in ids_cat:", missing_in_numerical)
print("elements in ids_cat but not in ids_num:", missing_in_categorical)

# Merge the two given numerical and categorical datasets based on their unique_ID.
merge = pd.merge(cat,num,on="unique_id",how="outer")
#Print the combined dataset
print(merge)
# print(merge[merge['unique_id'].isin(missing_in_categorical)])

elements in ids_num but not in ids_cat: set()
elements in ids_cat but not in ids_num: {208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335}
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
397     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
398     997629  0.0  0.0  normal 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [85]:
# Calculate the percentage of rows that contain atleast one missing value
num_of_miss = merge.isnull().any(axis=1).sum()
total = len(merge)
percent = num_of_miss / total
# Print %
print(f"percentage of missing rows with missing data {percent}")
# Drop these rows from the dataset
merge = merge.dropna()
# Print the Dataset
print(merge)

percentage of missing rows with missing data 0.6175
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118596  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
390     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
391     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
394     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent  

## Part 3.4: Sort the dataset according to the Labels

In [86]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merge = merge.sort_values(by='Target')
# Print the dataset
print(merge)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
103     337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
105     343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
108     349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
110     353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
171     486397  0.0  0.0    normal    normal  notpresent  notpresent   no   
43      203466  0.0  0.0    normal    normal  notpresent  notpresent   no   
169     482105  0.0  0.0    normal    normal  notpresent  notpresent   no   
54      228849  0.0  0.0    normal    normal  notpresent  notpresent   no   
396     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [87]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
merge = pd.get_dummies(merge,columns=categorical_columns)
# Print the dataset
print(merge)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
103     337834  4.0  0.0  48.0  70.0  117.0   56.0   3.8  111.0  2.5  ...   
105     343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
108     349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
110     353025  3.0  4.0  40.0  70.0  253.0  150.0  11.9  132.0  5.6  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
171     486397  0.0  0.0  20.0  70.0  123.0   44.0   1.0  135.0  3.8  ...   
43      203466  0.0  0.0  22.0  60.0   97.0   18.0   1.2  138.0  4.3  ...   
169     482105  0.0  0.0  44.0  60.0   95.0   46.0   0.5  138.0  4.2  ...   
54      228849  0.0  0.0  58.0  80.0  100.0   50.0   1.2  140.0  3.5  ...   
396     996076  0.0  0.0  57.0  60.0  105.0   49.0   1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.



Answer: This would be technicaly equivalent since for rbc column instead of having two columns for normla and 
abnormal, you could just have rbc where 0 = normal and 1 = abnormal. This would still be a binary result. However,
 if the categorical vairble could take more than 2 values one coudld not repersent it in a binary varible since a 
 binary can only repersent two states so we would need more then dumy column in order to repsent the result. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [88]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
mean = merge[numerical_columns].mean()
std = merge[numerical_columns].std()

rows_remove = []
for index, row in merge.iterrows():
    for col in numerical_columns:
        val = row[col]
        if val < (mean[col] - 3 * std[col]) or val > (mean[col] + 3 * std[col]):
            rows_remove.append(index)
            break
        
merge = merge.drop(index=rows_remove)
# Print the dataset
print(merge)

     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  ...  \
105     343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2  ...   
131     397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  ...   
149     438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  ...   
163     474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5  ...   
164     475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6  ...   
..         ...  ...  ...   ...    ...    ...    ...  ...    ...  ...  ...   
171     486397  0.0  0.0  20.0   70.0  123.0   44.0  1.0  135.0  3.8  ...   
43      203466  0.0  0.0  22.0   60.0   97.0   18.0  1.2  138.0  4.3  ...   
169     482105  0.0  0.0  44.0   60.0   95.0   46.0  0.5  138.0  4.2  ...   
54      228849  0.0  0.0  58.0   80.0  100.0   50.0  1.2  140.0  3.5  ...   
396     996076  0.0  0.0  57.0   60.0  105.0   49.0  1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [89]:
# Normalize the all Numerical Attributes in the dataset.
for col in numerical_columns:
    min = merge[col].min()
    max = merge[col].max()
    merge[col] = (merge[col] - min) / (max - min)
# Print the dataset
print(merge)

     unique_id   al   su       age    bp       bgr        bu        sc  \
105     343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
131     397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
149     438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
163     474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
164     475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
..         ...  ...  ...       ...   ...       ...       ...       ...   
171     486397  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074   
43      203466  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765   
169     482105  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346   
54      228849  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765   
396     996076  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765   

          sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
105  0.500000  0.657143  ...   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

We do not need the unqiue_id coloumn since it does not need to be used for any data maniipualtion and is just an idenitfier. 

Answer:

In [90]:
#Remove that column
merge = merge.drop(columns='unique_id')
# Print the dataset
print(merge)

      al   su       age    bp       bgr        bu        sc       sod  \
105  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
131  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
149  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
163  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
164  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..   ...  ...       ...   ...       ...       ...       ...       ...   
171  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
43   0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
169  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
54   0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
396  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
105  0.657143  0.172131  ...   False     Tr

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [91]:
# Export the dataframe to a new csv file
merge.to_csv("merged_data.csv", index=False)

# Export to JSON
merge.to_json("merged_data.json", orient="records", lines=True)


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 
| al | su | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|----------|------|---------------|-------------------|-------------------|------|--------------------|-------------------|-------------------|-------------------|-------------------|---------------|------------|--------------|-----------|-----------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|--------|-------------|----------------|
| 2.0 | 0.0 | 0.743243 | 0.5 | 0.442060 | 0.901961 | 0.432099 | 0.5 | 0.657143 | 0.172131 | 0.210526 | 0.395161 | 0.153846 | True | False | True | False | True | False | True | False | False | True | False | True | False | True | False | True | False | True | False | True | False | True |
| 2.0 | 2.0 | 0.770270 | 1.0 | 0.901288 | 0.163399 | 0.345679 | 0.767| 0.171429 | 0.606557 | 0.631579 | 0.443548 | 0.410256 | False | True | False | True | True | False | False | True | False | True | True | False | False | True | True | False | True | False | True | False | True | False |
| 2.0 | 0.0 | 0.905405 | 0.5 | 0.785408 | 0.862745 | 0.518519 | 0.6 | 0.828571 | 0.401639 | 0.447368 | 0.233871 | 0.435897 | True | False | True | False | True | False | True | False | False | True | False | True | False | True | False | True | False | True | False | True | False | True |
| 4.0 | 0.0 | 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.167| 0.171429 | 0.221311 | 0.184211 | 0.653226 | 0.333333 | False | True | True | False | False | True | False | True | True | False | True | False | True | False | True | False | True | False | True | False | True | False |
| 4.0 | 2.0 | 0.783784 | 1.0 | 0.399142 | 0.287582 | 0.839506 | 0.667| 0.485714 | 0.188525 | 0.263158 | 0.258065 | 0.205128 | True | False | True | False | True | False | False | True | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| 3.0 | 1.0 | 0.729730 | 0.0 | 0.935622 | 0.169935 | 0.160494 | 0.333| 0.028571 | 0.188525 | 0.236842 | 0.879032 | 0.102564 | False | True | True | False | False | True | True | False | False | True | True | False | True | False | True | False | True | False | True | False | False | True |
| 2.0 | 0.0 | 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.367| 0.542857 | 0.286885 | 0.342105 | 0.169355 | 0.205128 | True | False | True | False | True | False | True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False |
| 4.0 | 3.0 | 0.851352 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.0 | 0.285714 | 0.311475 | 0.315789 | 0.580645 | 0.179487 | False | True | True | False | False | True | False | True | False | True | False | True | False | True | True | False | False | True | True | False | True | False |
| 1.0 | 0.0 | 0.540541 | 0.0 | 0.399142 | 0.535948 | 0.358025 | 0.7 | 0.314286 | 0.344262 | 0.315789 | 0.830645 | 0.153846 | False | True | False | True | True | False | True | False | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| 3.0 | 0.0 | 0.756757 | 0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533| 0.514286 | 0.573770 | 0.605263 | 0.290323 | 0.333333 | False | True | True | False | True | False | True | False | False | True | False | True | False | True | False | True | False | True | True | False | True | False |
| 3.0 | 1.0 | 0.662162 | 0.5 | 0.618026 | 0.411765 | 0.432099 | 0.567| 0.571429 | 0.434426 | 0.473684 | 0.250000 | 0.282051 | False | True | True | False | False | True | False | True | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| 4.0 | 1.0 | 0.783784 | 0.0 | 0.725322 | 0.313725 | 0.481481 | 0.567| 0.714286 | 0.319672 | 0.342105 | 0.258065 | 0.205128 | True | False | True | False | True | False | False | True | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| 4.0 | 0.0 | 0.878378 | 0.0 | 0.206009 | 0.751634 | 0.604938 | 0.533| 0.571429 | 0.475410 | 0.500000 | 0.879032 | 0.435897 | False | True | False | True | True | False | True | False | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| 3.0 | 0.0 | 0.878378 | 0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433| 0.428571 | 0.393443 | 0.447368 | 0.104839 | 0.256410 | False | True | True | False | False | True | False | True | False | True | False | True | False | True | True | False | True |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

We could check manually using pandas for simliar transformatoins to make sure everything is right. We could check the data before and after for some key examples and see if the transfomration applied make sense, like we want it to add every 2 rows toghter, we expect x + y = z but might get a instead. Also could test on a small samlpe size ratheer then doing everything at once. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [92]:
# Code to rename all the columns in the dataset
merge.rename(columns={
    'age': 'Age(numerical)',
    'bp': 'Blood Pressure(numerical)',
    'al': 'Albumin(nominal)',
    'su': 'Sugar(nominal)',
    'bgr': 'Blood Glucose Random(numerical)',
    'bu': 'Blood Urea(numerical)',
    'sc': 'Serum Creatinine(numerical)',
    'sod': 'Sodium(numerical)',
    'pot': 'Potassium(numerical)',
    'hemo': 'Hemoglobin(numerical)',
    'pcv': 'Packed Cell Volume(numerical)',
    'wbcc': 'White Blood Cell Count(numerical)',
    'rbcc': 'Red Blood Cell Count(numerical)',
    'rbc_abnormal': 'Red Blood Cells Abnormal(nominal)',
    'rbc_normal': 'Red Blood Cells Normal(nominal)',
    'pc_abnormal': 'Pus Cell Abnormal(nominal)',
    'pc_normal': 'Pus Cell Normal(nominal)',
    'pcc_notpresent': 'Pus Cell Clumps Not Present(nominal)',
    'pcc_present': 'Pus Cell Clumps Present(nominal)',
    'ba_notpresent': 'Bacteria Not Present(nominal)',
    'ba_present': 'Bacteria Present(nominal)',
    'htn_no': 'Hypertension No(nominal)',
    'htn_yes': 'Hypertension Yes(nominal)',
    'dm_no': 'Diabetes Mellitus No(nominal)',
    'dm_yes': 'Diabetes Mellitus Yes(nominal)',
    'cad_no': 'Coronary Artery Disease No(nominal)',
    'cad_yes': 'Coronary Artery Disease Yes(nominal)',
    'appet_good': 'Appetite Good(nominal)',
    'appet_poor': 'Appetite Poor(nominal)',
    'pe_no': 'Pedal Edema No(nominal)',
    'pe_yes': 'Pedal Edema Yes(nominal)',
    'ane_no': 'Anemia No(nominal)',
    'ane_yes': 'Anemia Yes(nominal)',
    'Target_ckd': 'Class (CKD)(nominal)',
    'Target_notckd': 'Class (Not CKD)(nominal)'
}, inplace=True)

# Display the updated DataFrame
print(merge)

     Albumin(nominal)  Sugar(nominal)  Age(numerical)  \
105               2.0             0.0        0.743243   
131               2.0             2.0        0.770270   
149               2.0             0.0        0.905405   
163               4.0             0.0        0.202703   
164               4.0             2.0        0.783784   
..                ...             ...             ...   
171               0.0             0.0        0.189189   
43                0.0             0.0        0.216216   
169               0.0             0.0        0.513514   
54                0.0             0.0        0.702703   
396               0.0             0.0        0.689189   

     Blood Pressure(numerical)  Blood Glucose Random(numerical)  \
105                       0.50                         0.442060   
131                       1.00                         0.901288   
149                       0.50                         0.785408   
163                       0.75                 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [93]:
# Converted SQL to Pandas code
data = {
    'Target': ['A', 'B', 'A', 'C', 'B', 'A', 'C', 'C'],
    'OtherColumn': [1, 2, 3, 4, 5, 6, 7, 8]  # Just an example column
}

# Creating the DataFrame
df = pd.DataFrame(data)

# Grouping by 'Target' and counting occurrences
result = df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

  Target  count
0      A      3
1      B      2
2      C      3


*Describe what the above code does here*
The query counts the number of times the an example in the Target coulumn appers with in itself, using the Count to count the number of times a perticaulr data example appers and store that number in a another coulumn next to the given. 
